### PR TITLE
jupyter extensions

### PR DESCRIPTION
**Pull request by GaetanLepage**
###### Description of changes

Add the following python packages:

- jupyter-contrib-core
- jupyter-nbextensions-configurator
- jupyter-highlight-selected-word
- jupyter-contrib-nbextensions

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 243220` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>python310Packages.jupyter-contrib-core</li>
    <li>python310Packages.jupyter-contrib-core.dist</li>
    <li>python310Packages.jupyter-contrib-nbextensions</li>
    <li>python310Packages.jupyter-contrib-nbextensions.dist</li>
    <li>python310Packages.jupyter-highlight-selected-word</li>
    <li>python310Packages.jupyter-highlight-selected-word.dist</li>
    <li>python310Packages.jupyter-nbextensions-configurator</li>
    <li>python310Packages.jupyter-nbextensions-configurator.dist</li>
    <li>python311Packages.jupyter-contrib-core</li>
    <li>python311Packages.jupyter-contrib-core.dist</li>
    <li>python311Packages.jupyter-contrib-nbextensions</li>
    <li>python311Packages.jupyter-contrib-nbextensions.dist</li>
    <li>python311Packages.jupyter-highlight-selected-word</li>
    <li>python311Packages.jupyter-highlight-selected-word.dist</li>
    <li>python311Packages.jupyter-nbextensions-configurator</li>
    <li>python311Packages.jupyter-nbextensions-configurator.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/jupyter-contrib-core/default.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{ buildPythonPackage
+, fetchFromGitHub
+, jupyter-core
+, lib
+, notebook
```
Comment (figsoda) [inline pkgs/development/python-modules/jupyter-contrib-core/default.nix:5]: > [Arguments should be listed in the order they are used, with the exception of `lib`, which always goes first.](https://nixos.org/manual/nixpkgs/unstable/#sec-syntax)

```suggestion
{ lib
, buildPythonPackage
, fetchFromGitHub
, jupyter-core
, notebook
```

same suggestion for the other files

Comment (GaetanLepage) [reply pkgs/development/python-modules/jupyter-contrib-core/default.nix]: Ok ! I was not aware of this rule.
I fixed it on all 4 files, thanks !

**Comment by figsoda** [comment]
@ofborg build python3Packages.jupyter-contrib-core python3Packages.jupyter-nbextensions-configurator python3Packages.jupyter-highlight-selected-word 
python3Packages.jupyter-contrib-nbextensions

File: pkgs/development/python-modules/jupyter-contrib-nbextensions/default.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+, lxml
+}:
+
+buildPythonPackage rec {
+  pname = "jupyter-contrib-nbextensions";
+  version = "0.7.0";
+
+  src = fetchFromGitHub {
+    owner = "ipython-contrib";
+    repo = "jupyter-contrib-nbextensions";
```
Comment (figsoda) [inline pkgs/development/python-modules/jupyter-contrib-nbextensions/default.nix:16]:
```suggestion
    repo = "jupyter_contrib_nbextensions";
```

**Comment by figsoda** [comment]
btw ofborg will build the packages if you change the commit messages from `python3Packages` to `python310Packages`

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 243220` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>python310Packages.jupyter-contrib-core</li>
    <li>python310Packages.jupyter-contrib-core.dist</li>
    <li>python310Packages.jupyter-contrib-nbextensions</li>
    <li>python310Packages.jupyter-contrib-nbextensions.dist</li>
    <li>python310Packages.jupyter-highlight-selected-word</li>
    <li>python310Packages.jupyter-highlight-selected-word.dist</li>
    <li>python310Packages.jupyter-nbextensions-configurator</li>
    <li>python310Packages.jupyter-nbextensions-configurator.dist</li>
    <li>python311Packages.jupyter-contrib-core</li>
    <li>python311Packages.jupyter-contrib-core.dist</li>
    <li>python311Packages.jupyter-contrib-nbextensions</li>
    <li>python311Packages.jupyter-contrib-nbextensions.dist</li>
    <li>python311Packages.jupyter-highlight-selected-word</li>
    <li>python311Packages.jupyter-highlight-selected-word.dist</li>
    <li>python311Packages.jupyter-nbextensions-configurator</li>
    <li>python311Packages.jupyter-nbextensions-configurator.dist</li>
  </ul>
</details>

**Review by figsoda — APPROVED** [review]

File: pkgs/development/python-modules/jupyter-contrib-core/default.nix [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    repo = "jupyter_contrib_core";
+    rev = "refs/tags/${version}";
+    hash = "sha256-UTtK+aKxBFkqKuHrt1ox8vdHyFz/9HiKFl7U4UQcG88=";
+  };
+
+  propagatedBuildInputs = [
+    jupyter-core
+    notebook
+  ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/jupyter-contrib-core/default.nix:23]: We can't run the tests https://github.com/Jupyter-contrib/jupyter_contrib_core/tree/master/tests ?

File: pkgs/development/python-modules/jupyter-contrib-nbextensions/default.nix [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    hash = "sha256-1o8tBfRw6jNcKfNE7xXrQaEhx+KOv7mLOruvuMDtJ1Q=";
+  };
+
+  propagatedBuildInputs = [
+    jupyter-contrib-core
+    jupyter-highlight-selected-word
+    jupyter-nbextensions-configurator
+    lxml
+  ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/jupyter-contrib-nbextensions/default.nix:27]: same here https://github.com/ipython-contrib/jupyter_contrib_nbextensions/tree/master/tests

File: pkgs/development/python-modules/jupyter-highlight-selected-word/default.nix [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  version = "0.2.0";
+
+  src = fetchFromGitHub {
+    owner = "jcb91";
+    repo = "jupyter_highlight_selected_word";
+    rev = "refs/tags/${version}";
+    hash = "sha256-KgM//SIfES46uZySwNR4ZOcolnJORltvThsmEvxXoIs=";
+  };
+
+  pythonImportsCheck = [ "jupyter_highlight_selected_word" ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/jupyter-highlight-selected-word/default.nix:17]: The package seems to have no tests, we should disable them with doCheck = false

File: pkgs/development/python-modules/jupyter-nbextensions-configurator/default.nix [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+  src = fetchFromGitHub {
+    owner = "jupyter-contrib";
+    repo = "jupyter_nbextensions_configurator";
+    rev = "refs/tags/${version}";
+    hash = "sha256-ovKYHATRAC5a5qTMv32ohU2gJd15/fRKXa5HI0zGp/0=";
+  };
+
+  propagatedBuildInputs = [ jupyter-contrib-core ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/jupyter-nbextensions-configurator/default.nix:19]: https://github.com/Jupyter-contrib/jupyter_nbextensions_configurator/tree/master/tests

**Comment by GaetanLepage** [comment]
@SuperSandro2000 what you want to say is that tests are currently not detected and thus not ran, right ?
Do you want me to create a PR to explicitly set `doCheck = false` for all of them ?

**Comment by SuperSandro2000** [comment]
Then the tests are probably not included in the pypi download. Just fetch from github. If packages have no tests we likely won't notice when they break in a more subtle way because of dependency updates. All the packages that have tests upstream should run them if possible.

**Comment by GaetanLepage** [comment]
> Then the tests are probably not included in the pypi download. Just fetch from github. If packages have no tests we likely won't notice when they break in a more subtle way because of dependency updates. All the packages that have tests upstream should run them if possible.

I see, thanks for the clarification.
However in this case, we are using `fetchFromGitHub` and not `fetchPypi`.
Do you want have a change in mind for those specific packages ?

**Comment by SuperSandro2000** [comment]
If the package has tests and we add pytestCheckHook to nativeBuildInputs, those tests should be run. Otherwise we might need to add the location of the tests to pytestCheckFlags.

**Pull request merged**